xamarin/Xamarin.Forms
# Shell android dispose

### Description of Change ###
This PR takes a set of the changes from this PR https://github.com/xamarin/Xamarin.Forms/pull/6715 but in a slightly more subtle way plus it maintains the OnDestroy calls.

If we merge this version of the  PR then if we feel the other PR still has valid changes than we can rebase that one against master. I'd feel more comfortable with those changes living in master instead of just being pushed to 4.2.0. This PR also only has breaking changes with the removal of some constructors

I tested this PR with Xaminals and I'm not able to reproduce the back button bug


### Issues Resolved ### 
- fixes #6640
- closes #6715

### Platforms Affected ### 
- Android


### Testing Procedure ###
- use these nugets with xaminals and see if you can cause a dispose exception
- Run the control gallery and try to cause a dispose exception by swapping out the Shell Page or hitting back button until it exits CG

### PR Checklist ###
